orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.cc

```diff
@@ -0,0 +1,215 @@
+  }
+
+  auto squeeze_out_arg = &graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("squeeze_adaptor"),
+                                                   non_zero_out_arg->TypeAsProto());
+  Node& squeeze_node = graph.AddNode(graph.GenerateNodeName("squeeze_adaptor"), "Squeeze", "nonzero_squeezer",
+                                     squeeze_input_args, {squeeze_out_arg}, &attributes, kOnnxDomain);
+  ORT_ENFORCE(graph.SetOpSchemaFromRegistryForNode(squeeze_node),
+              "Failed to set op schema for " + squeeze_node.Name());
+
+  // After Squeeze, the shape becomes (valid_token_count).
+  ONNX_NAMESPACE::TensorShapeProto squeeze_output_shape;
+  squeeze_output_shape.add_dim()->set_dim_param(dim_name);
+  squeeze_out_arg->SetShape(squeeze_output_shape);
+  squeeze_node.SetExecutionProviderType(node.GetExecutionProviderType());
+
+  return squeeze_out_arg;
+}
+}  // namespace
+
+Status InsertGatherBeforeSceLoss::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/,
+                                            const logging::Logger& logger) const {
+  LOG_DEBUG_INFO(logger, "Enter InsertGatherBeforeSceLoss");
+
+  GraphViewer graph_viewer(graph);
+  size_t handled_sce_node_count = 0;  // For summary
+  const auto& order = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto index : order) {
+    auto* node_ptr = graph.GetNode(index);
+    if (!node_ptr)
+      // node was removed.
+      continue;
+
+    auto& node = *node_ptr;
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "SoftmaxCrossEntropyLossInternal", {1}, kMSDomain) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders())) {
+      continue;
+    }
+
+    // Found SoftmaxCrossEntropyLossInternal, check all requirements.
+
+    auto logits_shape = node.MutableInputDefs()[0]->Shape();
+    auto labels_shape = node.MutableInputDefs()[1]->Shape();
+    if (logits_shape == nullptr || labels_shape == nullptr) {
+      LOG_DEBUG_INFO(logger, "Skip SoftmaxCrossEntropyLossInternal node " + node.Name() +
+                                 " due to undefined input shapes.");
+      continue;
+    }
+
+    if (logits_shape->dim_size() != 2 || labels_shape->dim_size() != 1) {
+      LOG_DEBUG_INFO(logger, "Skip SoftmaxCrossEntropyLossInternal node " + node.Name() +
+                                 " due to unsupported input shape ranks.");
+      continue;
+    }
+
+    std::string reduction = node.GetAttributes().at("reduction").s();
+    bool need_reset_shape = false;
+    if (reduction.compare("mean") == 0 || reduction.compare("sum") == 0) {
+      // loss output is a scalar, don't need reset shape.
+    } else if (reduction.compare("none") == 0) {
+      need_reset_shape = true;
+    } else {
+      LOG_DEBUG_INFO(logger, "Skip SoftmaxCrossEntropyLossInternal node " + node.Name() +
+                                 " due to unsupported reduction type.");
+      continue;
+    }
+
+    if (need_reset_shape) {
```

Comment:
May we combine the if block with the `else if (reduction.compare("none") == 0)` block. We can change it in the future if we add support for it.